### PR TITLE
Workspace integration

### PR DESCRIPTION
**Pull request by erri120**
Part of #706.

- [x] add "add panel" button to toolbar
- [x] rebase and fix conflicts
- [x] resolve TODOs in code
  - [x] cleanup
  - [x] better API

### CODE REVIEW

**Comment by erri120** [comment]
```
System.ArgumentOutOfRangeException: Index was out of range. Must be non-negative and less than the size of the collection. (Parameter 'index')
   at System.Collections.Generic.List`1.get_Item(Int32 index)
   at Avalonia.Collections.AvaloniaList`1.get_Item(Int32 index)
   at Avalonia.Visual.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Layout.Layoutable.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Input.InputElement.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Controls.Control.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Visual.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Layout.Layoutable.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Input.InputElement.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Controls.Control.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Visual.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Layout.Layoutable.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Input.InputElement.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Controls.Control.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Visual.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Layout.Layoutable.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Input.InputElement.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Controls.Control.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Visual.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Layout.Layoutable.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Input.InputElement.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Controls.Control.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Visual.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Layout.Layoutable.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Input.InputElement.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Controls.Control.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Visual.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Layout.Layoutable.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Input.InputElement.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Controls.Control.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Visual.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Layout.Layoutable.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Input.InputElement.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Controls.Control.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Visual.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Layout.Layoutable.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Input.InputElement.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Controls.Control.OnDetachedFromVisualTreeCore(VisualTreeAttachmentEventArgs e)
   at Avalonia.Controls.TopLevel.HandleClosed()
   at Avalonia.Controls.WindowBase.HandleClosed()
   at Avalonia.Controls.Window.HandleClosed()
   at Avalonia.X11.X11Window.Cleanup(Boolean fromDestroyNotification)
   at Avalonia.X11.X11Window.Dispose()
   at Avalonia.Controls.Window.CloseInternal()
   at Avalonia.Controls.Window.HandleClosing(WindowCloseReason reason)
   at Avalonia.X11.X11Window.OnEvent(XEvent& ev)
   at Avalonia.X11.X11PlatformThreading.HandleX11(CancellationToken cancellationToken)
   at Avalonia.X11.X11PlatformThreading.RunLoop(CancellationToken cancellationToken)
   at Avalonia.Threading.DispatcherFrame.Run(IControlledDispatcherImpl impl)
   at Avalonia.Threading.Dispatcher.PushFrame(DispatcherFrame frame)
   at Avalonia.Threading.Dispatcher.MainLoop(CancellationToken cancellationToken)
   at Avalonia.Controls.ApplicationLifetimes.ClassicDesktopStyleApplicationLifetime.Start(String[] args)
   at Avalonia.ClassicDesktopStyleApplicationLifetimeExtensions.StartWithClassicDesktopLifetime(AppBuilder builder, String[] args, ShutdownMode shutdownMode)
   at NexusMods.App.Startup.Main(IServiceProvider provider, String[] args) in /ssd-pool/projects/NexusMods/app/main/src/NexusMods.App/Startup.cs:line 19
   at NexusMods.App.StartupHandler.StartUiWindowAsync() in /ssd-pool/projects/NexusMods/app/main/src/NexusMods.App/StartupHandler.cs:line 52
   at NexusMods.SingleProcess.StartupDirector.Start(String[] args, Boolean debugMode, IAnsiConsole console)
```

**Comment by erri120** [comment]
Opened https://github.com/AvaloniaUI/Avalonia/issues/14437

File: src/NexusMods.App.UI/Controls/Spine/SpineViewModel.cs [hunk 1]
```diff
@@ -127,85 +96,42 @@ private Bitmap LoadImageFromStream(Stream iconStream)
     private void NavigateToHome()
     {
         _logger.LogTrace("Home selected");
-        _actions.OnNext(new SpineButtonAction(Type.Home));
-        LeftMenu = _homeLeftMenuViewModel;
+        // TODO:
```
Comment (Al12rs) [inline src/NexusMods.App.UI/Controls/Spine/SpineViewModel.cs:99]: This and Downloads will be for another PR?

Comment (erri120) [reply src/NexusMods.App.UI/Controls/Spine/SpineViewModel.cs:99]: Yup

File: src/NexusMods.App.UI/RightContent/FoundGamesViewModel.cs [hunk 2]
```diff
@@ -33,7 +31,6 @@ private async Task ManageGame(GameInstallation installation)
     {
         var name = _loadoutRegistry.SuggestName(installation);
         var marker = await _loadoutRegistry.Manage(installation, name);
```
Comment (Al12rs) [inline src/NexusMods.App.UI/RightContent/FoundGamesViewModel.cs:33]: What's the plan here, move the navigation up or will it still happen here?

File: src/NexusMods.App.UI/LeftMenu/Game/GameLeftMenuViewModel.cs [hunk 3]
```diff
@@ -87,7 +85,9 @@ private void SelectLoadout(Loadout loadout)
     {
         _logger.LogDebug("Loadout {LoadoutId} selected", loadout.LoadoutId);
         _loadoutGridViewModel.LoadoutId = loadout.LoadoutId;
-        RightContent = _loadoutGridViewModel;
+
+        throw new NotImplementedException("Navigate to workspace");
```
[the remark this answers is not in the document]
Comment (erri120) [reply src/NexusMods.App.UI/LeftMenu/Game/GameLeftMenuViewModel.cs:89]: `GameLeftMenu*` can probably be removed since the spine doesn't contain individual games but individual loadouts. The original intend behind `GameLeftMenu*` was to select a game on the spine and then switch between loadouts in this menu.

File: src/NexusMods.App.UI/RightContent/FoundGamesViewModel.cs (hunk 2 above)
Comment (Al12rs) [reply src/NexusMods.App.UI/RightContent/FoundGamesViewModel.cs:33]: Currently the navigation here is problematic because it only happens after the game has bee completely indexed and backed up, while ideally it would switch immediately to a new loadout with the GameFiles mod entry loading.

Just a note for later

File: src/NexusMods.App.UI/Windows/MainWindowViewModel.cs [hunk 4]
```diff
@@ -88,32 +96,20 @@ public class MainWindowViewModel : AViewModel<IMainWindowViewModel>
...
-                }).BindTo(this, vm => vm.RightContent);
-
             // Only show the updater if the metrics opt-in has been shown before, so we don't spam the user.
             if (!metricsOptInViewModel.MaybeShow())
                 updaterViewModel.MaybeShow();
 
+            this.WhenAnyValue(vm => vm.IsActive)
+                .Where(isActive => isActive)
+                .Select(_ => WindowId)
+                .BindTo(_windowManager, manager => manager.ActiveWindowId)
```
Comment (Al12rs) [inline src/NexusMods.App.UI/Windows/MainWindowViewModel.cs:106]: What is the advantage of this syntax over `.Subscribe(_ => _windowManager.ActiveWindowId = WindowId)` ?

Comment (erri120) [reply src/NexusMods.App.UI/Windows/MainWindowViewModel.cs:106]: See UI coding guidelines, setting values using `Subscribe` should generally be avoided.

File: src/NexusMods.App.UI/Windows/IWindowManager.cs [hunk 5]
```diff
@@ -0,0 +1,45 @@
+using System.Collections.ObjectModel;
+using System.Diagnostics.CodeAnalysis;
+using JetBrains.Annotations;
+
+namespace NexusMods.App.UI.Windows;
+
+/// <summary>
+/// Represents a manager of all existing windows.
```
Comment (Al12rs) [inline src/NexusMods.App.UI/Windows/IWindowManager.cs:8]: Add notes to comment: 
- We support multiple mainwindows. 
- Purpose for this is to help to identify the current window.

File: src/NexusMods.App.UI/WorkspaceSystem/Controllers/WorkspaceController.cs [hunk 6]
```diff
@@ -1,39 +1,86 @@
...
 {
+    private readonly IWorkspaceWindow _window;
+    private readonly IServiceProvider _serviceProvider;
     private readonly ILogger<WorkspaceController> _logger;
-    private readonly Dictionary<WorkspaceId, WeakReference<WorkspaceViewModel>> _workspaces = new();
 
-    public WorkspaceController(ILogger<WorkspaceController> logger)
+    public WindowId WindowId => _window.WindowId;
+
+    private readonly SourceCache<WorkspaceViewModel, WorkspaceId> _workspaces = new(x => x.Id);
```
[the remark this answers is not in the document]
Comment (erri120) [reply src/NexusMods.App.UI/WorkspaceSystem/Controllers/WorkspaceController.cs:23]: `WorkspaceViewModel` has internal APIs that the controller uses.

[the remark this answers is not in the document]
Comment (Al12rs) [reply src/NexusMods.App.UI/WorkspaceSystem/Controllers/WorkspaceController.cs:23]: Ah the `AddPanel`, `OpenPage` etc are not in `IWorkspaceViewModel`, which is a good idea. 👍

File: src/NexusMods.App.UI/WorkspaceSystem/Page/PageFactoryController.cs [hunk 7]
```diff
@@ -43,12 +49,15 @@ public Page Create(PageData pageData, WorkspaceId workspaceId, PanelId panelId,
         return page;
     }
 
-    public IEnumerable<PageDiscoveryDetails> GetAllDetails()
+    /// <summary>
+    /// Returns all <see cref="PageDiscoveryDetails"/> of every factory given a <see cref="IWorkspaceContext"/>
+    /// </summary>
+    public IEnumerable<PageDiscoveryDetails> GetAllDetails(IWorkspaceContext workspaceContext)
```
Comment (Al12rs) [inline src/NexusMods.App.UI/WorkspaceSystem/Page/PageFactoryController.cs:55]: This is still not very clear, what are these `PageDiscoveryDetails`, as in what are they for?
We don't have a good explanation of the system and how it is supposed to be used.

Should we add a Doc page explaining how it all works? Or would you prefer to try to make things clearer by commenting the code more with details on purpose and how they are meant to be used?

**Review by Al12rs — CHANGES_REQUESTED** [review]

**Review by Al12rs — APPROVED** [review]
I still think we might need an overview doc to outline the general structure, but otherwise the PR is Ok.

I'm going through the other one as well
